snakemake/remote/zenodo.py

```diff
@@ -0,0 +1,212 @@
+import re
+import hashlib
+from collections import namedtuple
+
+from snakemake.remote import AbstractRemoteObject, AbstractRemoteProvider
+from snakemake.exceptions import ZenodoFileException, WorkflowError
+from snakemake.common import lazy_property
+from snakemake.utils import makedirs
+
+try:
+    # Third-party modules
+    import requests
+    from requests.exceptions import HTTPError
+except ImportError as e:
+    raise WorkflowError(
+        "The Python 3 package 'requests' "
+        + "must be installed to use Zenodo remote() file functionality. %s" % e.msg
+    )
+
+ZenFileInfo = namedtuple("ZenFileInfo", ["checksum", "filesize", "id", "download"])
+
+
+class RemoteProvider(AbstractRemoteProvider):
+    def __init__(self, *args, stay_on_remote=False, deposition=None, **kwargs):
+        super(RemoteProvider, self).__init__(
+            *args, stay_on_remote=stay_on_remote, deposition=deposition, **kwargs
+        )
+        self._zen = ZENHelper(*args, deposition=deposition, **kwargs)
+
+    def remote_interface(self):
+        return self._zen
+
+    @property
+    def default_protocol(self):
+        return "https://"
+
+    @property
+    def available_protocols(self):
+        return ["http://", "https://"]
+
+
+class RemoteObject(AbstractRemoteObject):
+    def __init__(
+        self,
+        *args,
+        keep_local=False,
+        stay_on_remote=False,
+        provider=None,
+        deposition=None,
+        **kwargs
+    ):
+        super(RemoteObject, self).__init__(
+            *args,
+            keep_local=keep_local,
+            stay_on_remote=stay_on_remote,
+            provider=provider,
+            deposition=deposition,
+            **kwargs
+        )
+        if provider:
+            self._zen = provider.remote_interface()
+        else:
+            self._zen = ZENHelper(*args, deposition=deposition, **kwargs)
+
+    # === Implementations of abstract class members ===
+    def _stats(self):
+        return self._zen.get_files()[os.path.basename(self.local_file())]
+
+    def exists(self):
+        return os.path.basename(self.local_file()) in self._zen.get_files()
+
+    def size(self):
+        if self.exists():
+            return self._stats().filesize
+        else:
+            return self._iofile.size_local
+
+    def mtime(self):
+        # There is no mtime info provided by Zenodo
+        # Hence, the files are always considered to be "ancient".
+        return 0
+
+    def download(self):
+        self._zen.download(self.local_file())
+
+    def upload(self):
+        if self.size() <= 100000000:
+            self._zen.upload(self.local_file(), self.remote_file())
+        else:
+            raise ZenodoFileException(
+                "Current Zenodo stable API supports <=100MB per file."
+            )
+
+    @property
+    def list(self):
+        return [i for i in self._zen.get_files()]
+
+    @property
+    def name(self):
+        return self.local_file()
+
+
+class ZENHelper(object):
```

Comment:
Wouldn't it make sense to use one of the zenodo helpers that are already there? For example pyzenodo?

Reply:
I considered using zenodo libraries, like Tommos0/pyzenodo and space-physics/pyzenodo3. I am not aware if there are any others. 
The first one didn't had API upload or download implemented, only searching for records which uses another api.
Pyzenodo3 looks very good, has also upload, but no download. 
So it seemed more straightforward to start from scratch and write another wrapper.
Plus, here we have zenodo sandbox for testing.